### PR TITLE
20 - AttributeError: Can't pickle local object 'augment_visit.<locals>.augment_func' 

### PR DESCRIPTION
**Pull request by dineshtrivedi**
Hello,

This is my attempt at fixing the AttributeError when trying to run pylint in parallel.
I have converted the closure methods/functions to be classes implementing __call__ so they can now be pickled. I then wrote a unit test to assert that. If you put the old code back the test fails.

I also copied part of the pre-commit and tox.ini configuration from [pylint](https://github.com/PyCQA/pylint) project. I hope that is fine

I still need to do a bit more testing, but I thought of starting the PR and getting feedback on the changes. I would really appreciate any feedback.

Main Changes:
- Replaced closure+function to keep state with classes implementing __call_
- Removed conditional support for python2
- Removed unused `code` variable as it seems to be dead code and there are no unit tests indicating what that code entails

Project level and development environment changes:
- Added pre-commit + Fixed all black and isort issues
- Added initial version for setup.cfg
- Added initial documentation for testing

### CODE REVIEW

**Comment by dineshtrivedi** [comment]
relates to #20

File: .travis.yml [hunk 1]
```diff
@@ -4,6 +4,7 @@ python:
   - "3.5"
   - "3.6"
   - "3.7"
+  - "3.8"
```
Comment (carlio) [inline .travis.yml:7]: The `setup.py` lists supported versions as 3.6 to 3.10 (which is good!) so the CI config should match.

File: pylint_plugin_utils/__init__.py [hunk 2]
```diff
@@ -1,64 +1,75 @@
...
-    called at any point to trigger the continuation of other checks, or not at all to
-    prevent any further checking.
+    Augmenting functions should accept a 'chain' function, which runs the checker
+    method and possibly any other augmentations, and secondly an Astroid node.
+    "chain()" can be called at any point to trigger the continuation of other
+    checks, or not at all to prevent any further checking.
     """
 
-    if sys.version_info[0] <= 2:
-        checker = get_checker(linter, checker_method.im_class)
```
Comment (carlio) [inline pylint_plugin_utils/__init__.py:43]: Good that this old v2 compatibility code is going :)

**Comment by carlio** [comment]
This all looks great :)  Thank you @dineshtrivedi .

Seems like you're still adding changes so I'll wait till you're finished but as long as it keeps API functionality used by `pylint-django` (which as far as I can tell it does) then I'll merge and make a new release.

**Comment by dineshtrivedi** [comment]
Thank you @carlio 

I have made the change related to the ci. I have added the other supported python version and changed the coverage command to use pytest-cov

However, it seems the build is failing in Travis - https://travis-ci.org/github/PyCQA/pylint-plugin-utils/requests
Unfortunately, I can't see why. I wonder if it is because of this message:
image

It seems we need to move to [www.travis-ci.com](https://www.travis-ci.com/)

Adding to that:
- The coveralls badge is also not finding the project. Not sure what to do about this
- Code health badge is also broken. Should I remove it?

Do you know how we can fix these issues?

**Pending: Proper test with pylint-django**

**Comment by carlio** [comment]
Hmm, seems like travis refuses to build now without a paid plan for PyCQA which I'm not in charge of. I don't have the option to enable pylint-plugin-utils on travis right now sadly.

I think the best idea would be to move to using github actions for now. If you want to try that go ahead, otherwise I'll try to sort it out in the coming days.

For coveralls, I've enabled it in coveralls (turns out it wasn't enabled!) but it won't work until coverage is submitted, which won't work until CI is working.

**Comment by carlio** [comment]
@dineshtrivedi  I just made a dummy CI file in the master branch using the defaults. I'm not sure how github actions works, I figured using the example might be enough to get it started. If you rebase/merge master into your branch and overwrite the `ci.yml` with the changes you made, that should kick things off I guess.

**Comment by carlio** [comment]
Wait ignore me I am stupid :-)  your CI is in your repo fork of course!

**Comment by dineshtrivedi** [comment]
I have copied part of the ci.yml from pylint project - https://github.com/dineshtrivedi/pylint-plugin-utils/actions
It seems to be working on my side.

And the changes seem to not have broken pylint-django when I tested.

I think for now I am done @carlio 
Any other changes I can make in another PR with fewer changes so it is easier to review and test.

Let me know if there is anything else I can assist with

**Comment by carlio** [comment]
Thanks again @dineshtrivedi . I will get back to this in a couple of days after Christmas but I don't expect any problems.

**Comment by dineshtrivedi** [comment]
Sure, @carlio. Thank you for that
Have a good festive season :)

**Pull request merged**